rslib/src/scheduler/fsrs/retention.rs

```diff
@@ -139,10 +143,14 @@ impl Collection {
         let learn_cost = {
             let revlogs_filter = revlogs
                 .iter()
-                .filter(|r| r.review_kind == RevlogReviewKind::Learning && r.button_chosen >= 1)
+                .filter(|r| {
+                    r.review_kind == RevlogReviewKind::Learning
+                        && r.button_chosen >= 1
+                        && r.taken_millis > 0
+                })
                 .map(|r| r.taken_millis);
-            if total_first > 0.0 {
-                revlogs_filter.sum::<u32>() as f64 / total_first / 1000.0
+            if let Ok(length) = revlogs_filter.try_len() {
```

Comment:
I don't think this will work, as Rust won't know the length without executing the iterator. You'll need revlogs_filter.clone().count().